---
title: "kSanity-VIRGO metagenomics counts QC"
author: Laura Vermeren & Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: refresh # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---


```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(gt)
library(patchwork)
library(SummarizedExperiment)
library(tidySummarizedExperiment)
library(labelled)
library(vegan)
library(RColorBrewer)

# tmp <- fs::dir_map("R scripts mg/", source)
tmp <- fs::dir_map("R scripts/", source)

theme_set(theme_light())
```

## Loading the data

```{r}

data_source <- "real"
data_dir <- get_data_dir(data_source)

if (data_source == "real") {
  mg_dir <- str_c(data_dir, "00 Raw/02 Metagenomics/")
  counts <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_ReadCounts_20250404.csv"))
  counts_corr <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_taxaGLcor_20250404.csv"))
  relabs <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_RelAbund_20250404.csv"))
  technical_metadata <- read_csv(str_c(mg_dir, "VIBRANT_MG_technicalMetaData_20250502.csv"))
  LBP_strain_info <- readxl::read_xlsx(str_c(data_dir, "00 Raw/00 Trial Data/IsolateNumbers.xlsx"))
  VIRGO2_taxonomic_table <- read_csv(str_c(mg_dir, "VIRGO2_taxonomy_key_250429.csv"))
} else {
  mg_dir <- str_c(data_dir, "03 metagenomics combined/")
  counts <- read_csv(str_c(mg_dir, "mg_combined.csv"))
  manifest <- read_csv(str_c(mg_dir, "mg_combined_manifest.csv"))
  
}

```

We load the following tables, all provided by Michael France:

- `counts` raw counts per LBP strain and taxa (dimensions: `r dim(counts) |> str_c(collapse = " samples x ") |> str_c(" columns")`)
- `counts_corr` counts per LBP strain and taxa corrected by gene length (dimensions: `r dim(counts_corr) |> str_c(collapse = " samples x ") |> str_c(" columns")`)
- `relabs` relative abundances per LBP strain and taxa, computed by M. France from `counts_corr` (dimensions: `r dim(relabs) |> str_c(collapse = " samples x ") |> str_c(" columns")`)
- `technical_metadata` technical metadata (dimensions: `r dim(technical_metadata) |> str_c(collapse = " samples x ") |> str_c(" columns")`)
- `VIRGO2_taxonomic_table` taxonomic table for the taxa as defined in VIRGO2 (dimensions: `r dim(VIRGO2_taxonomic_table) |> str_c(collapse = " taxa x ") |> str_c(" columns")`)

We note that some samples have been sequenced twice to increase the coverage. The technical metadata contains a column `LibrarySequencedTwice` that indicates whether a sample has been sequenced twice. 

```{r}

sample_sequences_twice <- 
  technical_metadata |> 
  group_by(UID) |> 
  mutate(n = n()) |> 
  ungroup() |> 
  arrange(-n) |> 
  filter(n > 1) 

```

There are `r nrow(sample_sequences_twice)/2` samples that have been sequenced twice.

```{r}
#| eval: false

sample_sequences_twice |> 
  select(UID, SampleType, Ext_Lib_Plate, SequencingRun, BioinformaticsProcessingBatch, `Selected4re-extraction`, LibrarySequencedTwice, Notes) |> 
  gt(caption = "Samples that have been sequenced twice") 

```

```{r}
rm(sample_sequences_twice)
```


> Note: once we have participants/visits info, we can check whether there are any systematic patterns in these samples.


Also, some samples have been flagged by Michael as "selected for re-extraction" (column `Selected4re-extraction`) in the manifest. 

```{r}
#| eval: false
technical_metadata |> filter(`Selected4re-extraction` == 1) |> 
  select(UID, SampleType, `Selected4re-extraction`, Notes) |> 
  left_join(
    counts |> select(sampleID, CST) |> 
      mutate(UID = sampleID)
  )

```



We note some negative (although extremely small) values in the `counts`, `counts_corr`, and `rel_abs` assays (all for native crispatus, I assume it being an artifact from kSANITY). We will set these to 0.


We also load `LBP_strain_info` from the VIBRANT Dropbox that contains information on the LBP strains

The names of the strains do not correspond exactly to those on the VIBRANT Dropbox folder. We modify them to match those provided by Michael.

```{r}

LBP_strain_info <- 
  LBP_strain_info |> 
  mutate(
    strain_id = `VMRC ID`, 
    LBP = ifelse(is.na(LC106), "LC-115", "LC-106 & LC-115") |> factor(), 
    strain_origin = `Geographic source` |> factor()
  ) |>
  dplyr::rename(Biose_ID = `Biose ID`) |> #VMRC_ID = `VMRC ID` 
  dplyr::select(strain_id, LBP, strain_origin, Biose_ID) |> 
  arrange(strain_origin, LBP) |> 
  mutate(strain_id = strain_id |> fct_inorder()) |> 
  dplyr::select(strain_id, LBP, strain_origin, contains("ID")) |> 
  mutate(
    strain_id_mg = sub("_.*", "", strain_id),
    strain_id_mg = 
      case_when(
        strain_id_mg == "CC0028A1" ~ "C0028A1", 
        TRUE ~ strain_id_mg
      )
  )

LBP_strain_info |>   
  set_variable_labels(
    strain_id = "Strain ID", 
    strain_id_mg = "Strain ID (as in metagenomics data)",
    strain_origin = "Strain origin", 
    Biose_ID = "Biose ID"
  ) |> 
  gt(caption = "LBP strain information") |> 
  tab_style(style = cell_text(weight = "bold"),
            locations = cells_column_labels())

```

## Creating a `SummarizedExperiment` object from these tables

The `SummarizedExperiment` object will contain the following assays:

- `counts`
- `counts_corr`
- `rel_abs`

The SE `colData` will contain the technical metadata, the total number of reads for each sample, along with the metagenomics-based CST, subCST, and VALENCIA's assignment scores. For samples that have been sequenced twice, we will merge the sequencing and processing information from both sequencing runs.

The SE `rowData` will contain the VIRGO2 taxonomic table, augmented with the LBP strain information.

> NOTE : At the moment, the table `counts` includes an extra column "multiGenera".
For now, we remove the `MuliGenera` from the `counts` data frame; but in the future, Michael should provide the corrected counts and relative abundances with this extra column.

```{r}

mg_to_SE <- function(counts, counts_corr, relabs, LBP_strain_info, technical_metadata, VIRGO2_taxonomic_table) {
  
  # First, we aggregate technical metadata by UID (e.g., if samples were sequenced twice)
  technical_metadata_uid <- 
    technical_metadata |> 
    group_by(UID) |> 
    arrange(DateSequenced) |> 
    summarise(
     # n = n(),
      across(everything(), ~ unique(.) |> na.omit() |> str_c(collapse = "; ")),
    ) |> 
    mutate(
      LibrarySequencedTwice = ifelse(LibrarySequencedTwice == 1, TRUE, FALSE),
      Ext_Lib_Plate = Ext_Lib_Plate |> as.integer()
    ) 
  
  # We also harmonize the participant ID
  technical_metadata_uid <- 
    technical_metadata_uid |> 
    mutate(
      participant = 
        case_when(
          str_detect(PID, "068-") ~ PID |> str_remove_all("-"),
          TRUE ~ PID
        )
    ) |> 
    relocate(participant, .after = PID) 
  
  # remove "multiGenera" from counts # TODO: change later
  counts <- 
    counts |> 
    select(-c(MultiGenera))

  assay_counts <- 
    counts |> 
    mutate(sample_id = sampleID |> str_remove("MG_")) |> 
    dplyr::select(-c(sampleID, CST, subCST, score)) |> 
    as.data.frame() |> 
    column_to_rownames("sample_id") |> 
    drop_na() |> 
    t() |> 
    pmax(0)
  
  assay_counts_corr <- 
    counts_corr |> 
    mutate(sample_id = sampleID |> str_remove("MG_")) |> 
    dplyr::select(-c(sampleID, CST, subCST, score)) |> 
    as.data.frame() |> 
    column_to_rownames("sample_id") |>
    t() |> 
    pmax(0)
  
  assay_relative_ab <- 
    relabs |> 
    mutate(sample_id = sampleID |> str_remove("MG_")) |> 
    dplyr::select(-c(sampleID, CST, subCST, score)) |> 
    as.data.frame() |> 
    column_to_rownames("sample_id") |>
    t() |> 
    pmax(0)
    
  
  # include a total reads per sample in the colData 
  se_coldata <-
    counts |> 
    mutate(sample_id = sampleID |> str_remove("MG_")) |> 
    select(sample_id, sampleID, CST, subCST, score) |>
    mutate(
      total_non_human_reads = rowSums(counts |> select(-c(sampleID, CST, subCST, score)))
      ) |> 
    select(sample_id, sampleID, total_non_human_reads, everything()) |> 
    left_join(
      technical_metadata_uid,
      by = c("sampleID" = "UID")
    ) |>
    as.data.frame()
  
  se_rowdata <- 
    tibble(
      taxon_id = 
        counts |> 
        dplyr::select(-c(sampleID, CST, subCST, score)) |>
        colnames()
    ) |> 
    left_join(
      VIRGO2_taxonomic_table |> 
        mutate(
          taxon_id = Taxa,
          last_available_taxonomic_level = 
            Full_taxonomy |> str_remove_all(".*;") |> str_remove_all("_.*")
        ) |> 
        relocate(Full_taxonomy, .after = Species),
      by = join_by(taxon_id)
    ) |> 
    left_join(
      LBP_strain_info |> 
        dplyr::rename(taxon_id = strain_id_mg),
      by = join_by(taxon_id)
    ) |> 
    mutate(
      taxon_label = 
        ifelse(!is.na(LBP), "LBP strain ", "") |> 
        str_c(taxon_id |> str_replace_all("_", " ")) |> 
        str_c(
          ifelse(
            last_available_taxonomic_level %in% c("g","f"), 
            str_c(" (", last_available_taxonomic_level, ")"), ""
          )
        ) |> 
        str_c(
          ifelse(taxon_id == "Lactobacillus_crispatus", " (native strains)", "")
        )
    ) |> 
    relocate(taxon_label, .after = taxon_id) |>
    as.data.frame() 
    
  # Harmonization of the order of samples and feature
  
  ordered_sample_ids <- se_coldata$sample_id
  assay_counts <- assay_counts[, ordered_sample_ids]
  assay_counts_corr <- assay_counts_corr[, ordered_sample_ids]
  assay_relative_ab <- assay_relative_ab[, ordered_sample_ids]

  ordered_taxa <- se_rowdata$taxon_id
  assay_counts <- assay_counts[ordered_taxa, ]
  assay_counts_corr <- assay_counts_corr[ordered_taxa, ]
  assay_relative_ab <- assay_relative_ab[ordered_taxa, ]

  SummarizedExperiment::SummarizedExperiment(
    assays = list(counts = assay_counts, counts_corr = assay_counts_corr, rel_abs = assay_relative_ab),
    rowData = se_rowdata,
    colData = se_coldata
  )
}

```

```{r}

SE_mg <- mg_to_SE(counts, counts_corr, relabs, LBP_strain_info, technical_metadata, VIRGO2_taxonomic_table)

```


## Checking the participant & visit codes

```{r}

SE_mg <- 
  SE_mg |> 
  mutate(
    location = ifelse(str_detect(participant, "06810"), "US", "SA")
  )

```


```{r}
#| fig-width: 14
#| fig-height: 5

SE_mg |> 
  colData() |> 
  as_tibble() |> 
  filter(SampleType == "ClinicalSample") |> 
  select(location, participant, VisitCode, VisitCodeFlagged) |> 
  group_by(participant, VisitCode) |> 
  mutate(n = n()) |> 
  ggplot() +
  aes(x = participant, y = VisitCode, fill = VisitCodeFlagged) +
  geom_tile() +
  geom_text(aes(label = n)) +
  scale_fill_manual(values = c("lightblue2", "red2")) +
  facet_grid(. ~ location, scales = "free", space = "free") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.position = "bottom"
    )

```

As noted by Michael (`VisitCodeFlagged`), some visit codes are unexpected. However, some visits flagged by Michael could be correct (0010 are "re-screening", 1001 and 1104 could be extra daily swabs (or "accidentally" shipped daily swabs)).

One participant (`98120009`) has weird visit codes (1010, 1020, 1030, ... , 1070) AND a weird participant ID (starting by 98 instead of 68)

In addition to the visit codes flagged by Michael, there is also a participant for whom there are two samples with different sample ID at the same visit:

```{r}

SE_mg |> 
  colData() |> 
  as_tibble() |> 
  filter(SampleType == "ClinicalSample") |> 
  select(sampleID, sample_id, location, participant, VisitCode) |> 
  group_by(participant, VisitCode) |> 
  mutate(n = n()) |> 
  ungroup() |> 
  filter(n > 1) |> 
  gt(caption = "Participant with two samples at the same visit") 

```



```{r}
#| fig-width: 14
#| fig-height: 5

tmp <- 
  SE_mg |> 
  colData() |> 
  as_tibble() |> 
  filter(SampleType == "ClinicalSample") |> 
  select(sample_id, location, participant, VisitCode, VisitCodeFlagged) |> 
  group_by(participant, VisitCode) |> 
  mutate(n = n()) |> 
  ungroup() |> 
  mutate(
    weird =
      case_when(
        (n > 1) | (VisitCode %in% seq(1010, 1070, by = 10)) | (VisitCode %in% c("0001", "0011")) ~ "Yes",
        (VisitCode %in% "0010") ~ "No",
        (VisitCodeFlagged == 1) ~ "Maybe",
        TRUE  ~ "No"
      ) |> 
    factor(levels = c("No", "Maybe", "Yes"))
  ) 

tmp |> 
  ggplot() +
  aes(x = participant, y = VisitCode, fill = weird) +
  geom_tile() +
  geom_text(aes(label = n)) +
  scale_fill_manual("Weird?", values = c("lightblue2","gold", "red1")) +
  facet_grid(. ~ location, scales = "free", space = "free") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.position = "bottom"
    )

```

```{r}

tmp |> filter(weird != "No") |> arrange(desc(weird)) |>  
  gt(caption = "List of suspect PID and VisitCode")

```



```{r}

tmp2 <- 
  SE_mg@colData |> 
  as.data.frame() |> 
  as_tibble() |> 
  left_join(tmp |> select(sample_id, weird), by = join_by(sample_id)) 

SE_mg$weird <- tmp2$weird

rm(tmp, tmp2)

```



## Exploratory & QC analyses

```{r}
SE_mg
```





### Total number of counts/relative abundances per sample

**Total number of non-human reads per sample**

```{r}

SE_mg |> 
  colData() |> 
  as_tibble() |>
  ggplot() +
  aes(x = total_non_human_reads, fill = LibrarySequencedTwice) +
  geom_vline(xintercept = 500000, linetype = "dashed", color = "black") +
  geom_histogram(bins = 50) +
  scale_x_log10() +
  facet_grid(SampleType ~ ., scale = "free") +
  labs(
    x = "Total number of non-human reads per sample", 
    y = "Number of samples"
  ) +
  theme(
    strip.text.y = element_text(angle = 0)
  )

```


```{r}

SE_mg |> 
  colData() |> 
  as_tibble() |>
  ggplot() +
  aes(x = total_non_human_reads, fill = `Selected4re.extraction`) +
  geom_vline(xintercept = 500000, linetype = "dashed", color = "black") +
  geom_histogram(bins = 50) +
  scale_x_log10() +
  facet_grid(SampleType ~ ., scale = "free") +
  labs(
    x = "Total number of non-human reads per sample", 
    y = "Number of samples"
  ) +
  theme(
    strip.text.y = element_text(angle = 0)
  )

```

```{r}

SE_mg |> 
  colData() |> 
  as_tibble()  |> 
  arrange(total_non_human_reads) |> 
  mutate(index = row_number()) |>
  select(index, sampleID, total_non_human_reads, SampleType, LibrarySequencedTwice, Selected4re.extraction, subCST) |> 
  filter(total_non_human_reads < 1e6) |>
  gt()

```


```{r}

SE_mg |> 
  colData() |> 
  as_tibble() |>
  filter(SampleType == "ClinicalSample", VisitCode %in% c(seq(1000, 1900, by = 100), 2120)) |> 
  ggplot() +
  aes(x = total_non_human_reads, fill = `Selected4re.extraction`) +
  geom_vline(xintercept = 500000, linetype = "dashed", color = "black") +
  geom_histogram(bins = 50) +
  scale_x_log10() +
  facet_grid(VisitCode ~ ., scale = "free") +
  labs(
    x = "Total number of non-human reads per sample", 
    y = "Number of samples"
  ) +
  theme(
    strip.text.y = element_text(angle = 0)
  )

```



**Total relative abundances per sample**

We check that the relative abundances sum to 1 for each sample.

```{r}
#| fig-height: 3
#| fig-width: 5

tol <- 1e-5

SE_mg |> 
  as_tibble() |> 
  group_by(.sample) |>
  summarise(total_rel_abs = sum(rel_abs)) |> 
  ggplot(aes(x = total_rel_abs)) +
  geom_histogram(binwidth = tol) + 
  labs(x = "Total relative abundances per sample", 
       y = "Number of samples") +
  scale_x_continuous(limits = 1 + c(-1, 1) * 5 * tol) 

rm(tol)

```



```{r}

# We check the relative abundances computation (should be `counts_corr/sum(counts_corr)`).

tmp <- 
  SE_mg |> 
  as_tibble() |> 
  group_by(.sample) |>
  mutate(rel_abs_check = counts_corr/sum(counts_corr)) |> 
  ungroup() 

if (max(tmp$rel_abs_check - tmp$rel_abs) > 0.01) warning("!! Relative abundances do not match the computed values !!")

```




### Non-bacterial DNA

There are some non-bacterial taxa in the dataset:

```{r}

SE_mg |> 
  as_tibble() |> 
  filter(Category != "Bacteria") |> 
  group_by(.sample, Category) |> 
  summarize(rel_abs = sum(rel_abs), .groups = "drop") |> 
  ggplot() +
  aes(x = rel_abs, fill = Category) +
  geom_histogram(binwidth = 0.001) +
  facet_wrap(~ Category, scales = "free") +
  scale_x_continuous("Total relative abundance per sample for that group", labels = scales::percent_format()) +
  ylab("Number of samples") +
  guides(fill = "none")
  
  
SE_mg |> 
  as_tibble() |> 
  filter(Category != "Bacteria") |> 
  group_by(.sample) |>
  mutate(total_rel_ab = sum(rel_abs)) |>
  ungroup() |> 
  arrange(-total_rel_ab, -rel_abs) |> 
  mutate(
    .sample = .sample |> fct_inorder(),
    taxon_label = taxon_label |> fct_inorder()
  ) |> 
  filter(as.numeric(.sample) <= 20) |>
  ggplot() +
  aes(x = rel_abs, y = .sample |> fct_rev(), fill = Category) +
  geom_col() +
  scale_x_continuous("Relative abundance", labels = scales::percent_format()) +
  ylab("Top 20 samples with most non-bacterial relative abundances") 


SE_mg |> 
  as_tibble() |> 
  filter(Category != "Bacteria") |> 
  group_by(taxon_label) |>
  mutate(total_rel_ab = sum(rel_abs), max_rel_ab = max(rel_abs)) |>
  ungroup() |> 
  arrange(-total_rel_ab, -rel_abs) |> 
  mutate(
    .sample = .sample |> fct_inorder(),
    taxon_label = taxon_label |> fct_inorder()
  ) |> 
  filter(as.numeric(taxon_label) <= 20, max_rel_ab > 0.001) |>
  ggplot() +
  aes(x = rel_abs, y = taxon_label |> fct_rev(), color = Category) +
  geom_point(alpha = 0.3) +
  scale_x_continuous("Relative abundance", labels = scales::percent_format()) +
  ylab("Top non-bacterial organisms") +
  theme(strip.text.y = element_text(angle = 0)) +
  labs(caption = "Each dot is a sample.")
  
  

```


Consequently, we create a new assay which contains the relative abundances of the bacteria only. 

```{r}

SE_mg <- 
  SE_mg |>
  mutate(
    rel_abs_bact = rel_abs * (!is.na(Category) & (Category == "Bacteria"))
  )

assay(SE_mg, "rel_abs_bact") <- t(t(assay(SE_mg, "rel_abs_bact"))/colSums(assay(SE_mg, "rel_abs_bact")))


```

### Control samples

There are 4 categories of control samples:

```{r}

SE_mg |> 
  colData() |> 
  as_tibble() |> 
  filter(SampleType != "ClinicalSample") |>
  select(SampleType, sampleID, Ext_Lib_Plate) |> 
  group_by(SampleType) |> 
  arrange(SampleType, Ext_Lib_Plate) |>
  gt(caption = "Control samples", row_group_as_column = TRUE) 

```


These control samples have been created such that

- Mock 1: pooled patient samples from CAP084 swabs

    + 50 non-BV
        - VMRC 15 L.crispatus isolates
        - 12 L.crispatus dominant samples
        - 23 L. iners dominant samples
        
    + 50 BV
        - 24 G. vaginalis dominant samples
        - BV bacteria, not G. vaginalis dominant
 
- Mock 2: Pooled pure isolates (equal volumes of pure isolates at OD 0.1)

    + VMRC 15 L.crispatus isolates
    + G. vaginalis ATCC
    + A. vaginae ATCC
    + P. bivia ATCC
    + L. crispatus ATCC
    + L. gasseri ATCC
    + S. agalactiae ATCC
    + L. jensenii ATCC
    + F. magna 


Taxonomic composition of these control samples:


```{r}

get_taxa_colors <- function(taxa) {
  tibble(
    taxa = taxa
  ) |> 
    mutate(
      cat = 
        case_when(
          taxa == "Other" ~ "Other",
          str_detect(taxa, "crispatus") ~ "crispatus",
          str_detect(taxa, "LBP") ~ "LBP",
          str_detect(taxa, "iner") ~ "iners",
          str_detect(taxa, "revotella") ~ "prevotella",
          str_detect(taxa, "ardnerella") ~ "gardnerella",
          TRUE ~ "non lacto"
        )
    ) |> 
    group_by(cat) |>
    mutate(
      color = 
        case_when(
          taxa == "Other" ~ "gray80",
          str_detect(taxa, "crispatus") ~ "orange",
          str_detect(taxa, "LBP") ~ colorRampPalette(c("orangered1", "red4"))(n()),
          str_detect(taxa, "iner") ~ "green3",
           str_detect(taxa, "jensenii") ~ "green4",
          str_detect(taxa, "ardnerella") ~ colorRampPalette(c("dodgerblue1", "dodgerblue4"))(n()),
          str_detect(taxa, "revotella") ~ colorRampPalette(c("purple1", "purple4"))(n()),
          TRUE ~ colorRampPalette(c("turquoise3", "black"))(n())
        )
    ) |> 
    pull(color)
}

```


```{r}

control_types <- SE_mg$SampleType |> unique() |> sort() |> extract(-1)

```

```{r}
#| fig-width: 8
#| fig-height: 6

map(
  control_types,
  ~ {
    tmp <- 
      SE_mg |> 
      as_tibble() |> 
      filter(SampleType == .x) |> 
      filter(rel_abs > 0) 
    top_taxa <- 
      tmp |> 
      group_by(taxon_label) |> 
      summarise(total_relabs = sum(rel_abs)) |> 
      arrange(desc(total_relabs)) |> 
      slice_head(n = 25) |> 
      pull(taxon_label) |> sort()
    
    all_taxa <- tmp$taxon_label |> unique() |> sort()
    
    tmp |> 
      mutate(
        sample_label = str_c(.sample, ifelse(`Selected4re-extraction` == 1, " (re-x)", ""))
      ) |> 
      ggplot() +
      aes(y = rel_abs, x = sample_label, fill = taxon_label) +
      geom_col(color = "white", linewidth = 0.1) +
      scale_x_discrete("Samples") +
      scale_y_continuous("Rel. ab.") +
      scale_fill_discrete(
        "Top taxa", breaks = top_taxa #, limits = all_taxa, values = get_taxa_colors(all_taxa)
        ) + 
      facet_grid(. ~ str_c("Plate ", Ext_Lib_Plate), scales = "free") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      labs(title = str_c("Control samples : ", .x))
  }
)

```


And it looks like the replicability across controls (plates?) is not very good.


#### Mock 1

We see that the relative abundances are not very replicable, but let's check if detection (presence/absence) is better.


```{r}

tmp <- 
  SE_mg |> 
  as_tibble() |> 
  filter(SampleType == "Mock 1") |> 
  mutate(present = (rel_abs > 1/1000)) |> 
  group_by(taxon_label) |>
  mutate(tot_prop = sum(rel_abs)) |> 
  ungroup() |> 
  arrange(-tot_prop) |> 
  mutate(taxon_label = taxon_label |> fct_inorder()) 

```

```{r}
#| fig-width: 10
#| fig-height: 4

tmp |> 
  ggplot() +
  aes(x = taxon_label, y = .sample, fill = rel_abs |> log10()) +
  geom_tile() +
  ylab("") +
  scale_x_discrete("Taxa", breaks = NULL) +
  scale_fill_gradient(low = "gray99", high = "red", na.value = "white")
```


#### Mock 2

In the Mock 2, we check for the proportion of species/taxa that were not expected based on the theoretical composition of Mock 2

```{r}

# Focus on Mock2

tmp <- 
  SE_mg |> 
  as_tibble() |> 
  filter(SampleType == "Mock 2") |> 
  filter(rel_abs > 0) |> 
  mutate(
    expected = 
      str_detect(taxon_label, "LBP") | 
      str_detect(taxon_label, "Gardnerella") |
      str_detect(taxon_label, "Fannyhessea") |
      str_detect(taxon_label, "bivia") |
      str_detect(taxon_label, "crispatus") |
      str_detect(taxon_label, "gasseri") |
      str_detect(taxon_label, "jensenii") |
      str_detect(taxon_label, "Streptococcus agalactiae") |
      str_detect(taxon_label, "Finegoldia magna") 
  )

tmp |> 
  ggplot() +
  aes(y = rel_abs, x = .sample, fill = expected) +
  geom_col(color = "white", linewidth = 0.1) +
  scale_x_discrete("Samples") +
  scale_y_continuous("Rel. ab.") + 
  facet_grid(. ~ str_c("Plate ", Ext_Lib_Plate), scales = "free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

tmp |> 
  group_by(.sample, Ext_Lib_Plate, expected) |> 
  summarize(tot_rel_ab_pc = round(100 * sum(rel_abs), 2), .groups = "drop") |>
  filter(!expected) |> 
  gt(caption = "Percent of relative abundance with unexpected taxa")
  

```

When only considering bacterial relative abundances, results look similar: 

```{r}

# Focus on Mock2

tmp <- 
  SE_mg |> 
  as_tibble() |> 
  filter(SampleType == "Mock 2") |> 
  filter(rel_abs_bact > 0) |> 
  mutate(
    expected = 
      str_detect(taxon_label, "LBP") | 
      str_detect(taxon_label, "Gardnerella") |
      str_detect(taxon_label, "Fannyhessea") |
      str_detect(taxon_label, "bivia") |
      str_detect(taxon_label, "crispatus") |
      str_detect(taxon_label, "gasseri") |
      str_detect(taxon_label, "jensenii") |
      str_detect(taxon_label, "Streptococcus agalactiae") |
      str_detect(taxon_label, "Finegoldia magna") 
  )

tmp |> 
  ggplot() +
  aes(y = rel_abs_bact, x = .sample, fill = expected) +
  geom_col(color = "white", linewidth = 0.1) +
  scale_x_discrete("Samples") +
  scale_y_continuous("Rel. ab. (of bacterial content)") + 
  facet_grid(. ~ str_c("Plate ", Ext_Lib_Plate), scales = "free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

tmp |> 
  group_by(.sample, Ext_Lib_Plate, expected) |> 
  summarize(tot_rel_ab_pc = round(100 * sum(rel_abs_bact), 2), .groups = "drop") |>
  filter(!expected) |> 
  gt(caption = "Percent of relative abundance with unexpected taxa")
  

```


It could be annoying that we are so close to the detection threshold for the primary outcomes definition.





#### Alpha-diversity of clinical samples and controls


```{r}

vegan::diversity(SE_mg |> assay("rel_abs"), index = "shannon", MARGIN = 2) |> 
  as_tibble() |> 
  mutate(sample_id = colnames(SE_mg)) |> 
  left_join(SE_mg |> colData() |> as_tibble(), by = join_by(sample_id)) |> 
  ggplot() +
  aes(x = SampleType, y = value |> exp(), col = SampleType, fill = SampleType) +
  geom_violin(scale = "width", color = NA, alpha = 0.2) +
  ggbeeswarm::geom_quasirandom(size = 0.5, alpha = 0.5) +
  scale_y_log10() +
  labs(
    x = "",
    y = "Effective number of species based on shannon diversity index"
  ) +
  guides(fill = "none", color = "none") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```


Negative controls have lower total reads and lower alpha-diversity. 



### Proportion of LBP strains per sample

Do we detect any LBP strain?

```{r}

SE_mg |> 
  as_tibble() |> 
  filter(!is.na(LBP)) |> 
  ggplot() +
  aes(x = rel_abs, fill = (rel_abs == 0)) +
  geom_histogram(binwidth = 0.01) +
  scale_fill_manual("", values = c("steelblue", "gray"), labels = c("Rel. ab = 0", "Rel. ab > 0")) +
  xlab("Relative abundance of LBP strains") +
  ylab("Number of samples x LBP strain")

```


```{r}

SE_mg |> 
  as_tibble() |> 
  filter(!is.na(LBP), rel_abs > 0) |> 
  ggplot() +
  aes(x = rel_abs) +
  geom_histogram(binwidth = 0.001) +
  xlab("Non-zero relative abundance of LBP strains") +
  ylab("Number of samples") +
  facet_grid(LBP + .feature ~ .) +
  theme(
    strip.text.y = element_text(angle = 0)
  )
  

```


```{r}
#| fig-width: 8
#| fig-height: 4

SE_mg |> 
  as_tibble() |> 
  filter(!is.na(LBP), rel_abs > 0) |>
  ggplot() +
  aes(x = .feature) +
  geom_bar() +
  scale_x_discrete("LBP strains") +
  scale_y_continuous("Number of samples with non-zero relative abundances") +
  facet_grid(. ~ LBP, scales = "free", space = "free") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
  
```


```{r}

number_of_non_zero_LBP_strains_per_sample <- 
  SE_mg |> 
  as_tibble() |>  
  filter(!is.na(LBP)) |> 
  group_by(.sample) |> 
  summarize(
    number_of_non_zero_LBP_strains = sum(rel_abs > 0),
  ) 

number_of_non_zero_LBP_strains_per_sample |> 
  ggplot() +
  aes(x = number_of_non_zero_LBP_strains) +
  geom_histogram(binwidth = 0.5) +
  scale_x_continuous("Number of non-zero LBP strains per sample", breaks = 0:15, minor_breaks = 0) +
  ylab("Number of samples") 

```

There are `r nrow(number_of_non_zero_LBP_strains_per_sample |> filter(number_of_non_zero_LBP_strains > 0))` samples (`r (nrow(number_of_non_zero_LBP_strains_per_sample |> filter(number_of_non_zero_LBP_strains > 0))/nrow(number_of_non_zero_LBP_strains_per_sample)*100) |> round(2)`%) with at least one LBP strain detected (non-zero relative abundance).


```{r}

SE_mg |> 
  as_tibble() |>  
  filter(!is.na(LBP)) |> 
  left_join(
    number_of_non_zero_LBP_strains_per_sample,
    by = join_by(.sample)
  ) |>
  filter(number_of_non_zero_LBP_strains != 0) |>
  group_by(.feature) |>
  mutate(tot_rel_abs_for_strain = sum(rel_abs)) |> 
  ungroup() |> 
  arrange(LBP, -tot_rel_abs_for_strain) |>
  mutate(.feature = .feature |> fct_inorder()) |> 
  group_by(.sample) |> 
  mutate(
    LBP_rel_abs = rel_abs / sum(rel_abs),
    score = weighted.mean(LBP_rel_abs, .feature |> as.numeric()),
    total_LBP_rel_abs = sum(rel_abs)
  ) |> 
  ungroup() |> 
  arrange(score) |> 
  mutate(.sample = .sample |> fct_inorder()) |> 
  ggplot() +
  aes(x = .feature, y = .sample, fill = rel_abs) +
  geom_tile() +
  scale_fill_continuous("Rel. ab", low = "white", high = "steelblue2") +
  scale_x_discrete("Strains") +
  scale_y_discrete("Samples", breaks = NULL) +
  facet_grid(. ~ LBP + strain_origin, scales = "free_x", space = "free") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(
    caption = "Rel. ab. of LBP strains in samples with at least one LBP strain detected (non-zero)",
  )

```



```{r}
#| eval: false

samples_with_only_one_LBP_strain <- 
  number_of_non_zero_LBP_strains_per_sample |> 
  filter(number_of_non_zero_LBP_strains == 1) |> 
  pull(.sample)

SE_mg |> 
  as_tibble() |> 
  filter(!is.na(LBP)) |> 
  left_join(number_of_non_zero_LBP_strains_per_sample, by = ".sample") |>
  filter(number_of_non_zero_LBP_strains != 0) |> 
  ggplot() +
  aes(x = rel_abs, fill = number_of_non_zero_LBP_strains |> factor()) +
  geom_histogram() +
  facet_grid(number_of_non_zero_LBP_strains ~ taxon_label) +
  theme(strip.text.y = element_text(angle = 0))

```

### Compositions

```{r}

# creating a taxa category "other"
summarized_rel_abs <- 
  SE_mg |> 
  as_tibble() |> 
  group_by(.feature) |> 
  mutate(max_rel_ab = max(rel_abs), mean_rel_ab = mean(rel_abs)) |> 
  ungroup() |> 
  mutate(
    taxa = 
      case_when(
        !is.na(LBP) ~ taxon_label,
        max_rel_ab > 1/3 ~ taxon_label,
        TRUE ~ "Other"
      ),
    is_lacto = str_detect(Genus, "Lactobacillus") & (taxa != "Other")
  ) |> 
  arrange(LBP, is_lacto, desc(max_rel_ab)) |>
  mutate(
    taxa = taxa |> fct_inorder()
  ) |> 
  group_by(.sample, SampleType, participant, VisitCode, weird, taxa, LBP, is_lacto) |>
  summarize(rel_abs = sum(rel_abs), .groups = "drop") |>
  arrange(.sample, -rel_abs) |> 
  group_by(.sample) |>
  mutate(
    score = weighted.mean(taxa |> as.numeric(), rel_abs), 
    tot = sum(rel_abs),
    dom = taxa[1]
    ) |> 
  ungroup() |> 
  arrange(score) |> 
  mutate(.sample = .sample |> fct_inorder())

```

```{r}
#| fig-width: 21
#| fig-height: 6

summarized_rel_abs |> 
  ggplot() +
  aes(x = .sample, y = rel_abs, fill = taxa) +
  geom_col() +
  scale_x_discrete("Samples", breaks = NULL) +
  facet_grid(. ~ SampleType, scales = "free", space = "free") +
  scale_fill_manual(
    "", 
    values = get_taxa_colors(summarized_rel_abs$taxa |> levels()),
    guide = guide_legend(nrow = 5)
    ) +
  theme(
    strip.text.x = element_text(angle = 90, hjust = 0),
    legend.position = "bottom",
  )

```


```{r}
#| fig-width: 12
#| fig-height: 8

summarized_rel_abs |> 
  mutate(location = ifelse(str_detect(participant, "06810"),"US", "SA")) |> 
  filter(SampleType == "ClinicalSample", VisitCode %in% c(seq(1000, 1900, by = 100), 2120), weird == "No") |>
  ggplot() +
  aes(x = participant, y = rel_abs, fill = taxa) +
  geom_col() +
  scale_x_discrete("Participant") + # , breaks = NULL) +
  facet_grid(VisitCode ~ location, scales = "free_x", space = "free_x") +
  scale_fill_manual(
    "", 
    values = get_taxa_colors(summarized_rel_abs$taxa |> levels()),
    guide = guide_legend(nrow = 5)
    ) +
  theme(
    strip.text.x = element_text(angle = 0, hjust = 0.5),
    strip.text.y = element_text(angle = 0),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.position = "bottom",
  )

```






### PCoA on counts

```{r}

dist_matrix <-
  vegdist(
    assay(SE_mg, "counts") |> t(), 
    method = "bray")

pcoa <- wcmdscale(dist_matrix, eig = TRUE)
print(pcoa)

pcoa$eig |>
  as.data.frame() |>
  rownames_to_column("axis") |>
  mutate(axis = str_remove(axis, "Eigenvalue"), 
         var_explained = 100*pcoa$eig/sum(pcoa$eig)) |>
  slice_head(n = 10) |>
  gt() |>
  tab_header(title = "Eigenvalues of the PCoA") |>
  cols_label(axis = "Axis", `pcoa$eig` = "Eigenvalue", var_explained = "Variance explained") |>
  fmt_number(columns = vars(`pcoa$eig`, var_explained), decimals = 3)

pcoa$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x = "Axis",
    y = "Eigenvalue"
  ) 


pcoa$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  filter(axis <= 20) |> 
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x = "Axis",
    y = "Eigenvalue"
  ) 

pcoa_data <- 
  tibble(
    sample_id = rownames(pcoa$points),
    PCoA = pcoa$points
  ) |> 
  left_join(
    SE_mg |> colData() |> as_tibble(),
    by = join_by(sample_id)
  )

```

```{r}

pcoa_plot <- function(pcoa_tb, pcoa, color_var, axes = 1:2){
  
  pcoa_tb <- 
    pcoa_tb |> 
    mutate(
      col = as.factor(!!sym(color_var)),
      x = PCoA[, axes[1]],
      y = PCoA[, axes[2]]
      )

  pcoa_tb |>
    ggplot() +
    aes(x = x, y = y, color = col) +
    geom_point(size = 2, alpha = 0.7) +
    coord_fixed() +
    labs(
      x = paste0("PCoA ", axes[1], " (", round(100 * pcoa$eig[axes[1]] / sum(pcoa$eig), 1), "%)"),
      y = paste0("PCoA ", axes[2], " (", round(100 * pcoa$eig[axes[2]] / sum(pcoa$eig), 1), "%)"),
      color = color_var
    ) 
}


```

#### Type of sample

```{r}

sample_type_colors <- c("gray", "dodgerblue", "orange", "black", "red")

pcoa_plot(pcoa_data, pcoa, "SampleType", c(1, 2)) + scale_color_manual(values = sample_type_colors) 
pcoa_plot(pcoa_data, pcoa, "SampleType", c(3, 4)) + scale_color_manual(values = sample_type_colors) 

```

```{r}
#| eval: false

manova <- adonis2(dist_matrix ~ pcoa_data$SampleType)
manova

```


#### Extraction Plate


```{r}

pcoa_plot(pcoa_data, pcoa, "Ext_Lib_Plate", c(1, 2))
pcoa_plot(pcoa_data, pcoa, "Ext_Lib_Plate", c(3, 4))

```


Permutation test 

```{r}
#| eval: false

manova <- adonis2(dist_matrix ~ pcoa_data$Ext_Lib_Plate)
manova

```



#### Selected for re-extraction

```{r}

pcoa_plot(pcoa_data, pcoa, "Selected4re.extraction", c(1, 2)) 
pcoa_plot(pcoa_data, pcoa, "Selected4re.extraction", c(3, 4)) 

```



#### Bioinformatics Processing Batch 

```{r}

pcoa_plot(pcoa_data, pcoa, "BioinformaticsProcessingBatch", c(1, 2))
pcoa_plot(pcoa_data, pcoa, "BioinformaticsProcessingBatch", c(3, 4))

```

#### Library Pool

```{r}

pcoa_plot(pcoa_data, pcoa, "Library.Pool", c(1, 2))
pcoa_plot(pcoa_data, pcoa, "Library.Pool", c(3, 4))

```


#### Lane


```{r}

pcoa_plot(pcoa_data |> filter(Lane %in% 1:4), pcoa, "Lane", c(1, 2))
pcoa_plot(pcoa_data |> filter(Lane %in% 1:4), pcoa, "Lane", c(3, 4))

```


### PCoA on relative abundances

```{r}

dist_matrix_relab <-
  vegdist(
    relabs |> dplyr::select(-c(CST, subCST, score)) |> column_to_rownames("sampleID"), 
    method = "bray")

pcoa_relab <- wcmdscale(dist_matrix_relab, eig = TRUE)
print(pcoa_relab)

pcoa_relab$eig |>
  as.data.frame() |>
  rownames_to_column("axis") |>
  mutate(axis = str_remove(axis, "Eigenvalue"), 
         var_explained = 100*pcoa_relab$eig/sum(pcoa_relab$eig)) |>
  slice_head(n = 10) |>
  gt() |>
  tab_header(title = "Eigenvalues of the PCoA") |>
  cols_label(axis = "Axis", `pcoa_relab$eig` = "Eigenvalue", var_explained = "Variance explained") |>
  fmt_number(columns = vars(`pcoa_relab$eig`, var_explained), decimals = 3)

pcoa_relab$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x= "Axis",
    y = "Eigenvalue"
  ) 


pcoa_relab$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  filter(axis <= 20) |> 
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x= "Axis",
    y = "Eigenvalue"
  ) 

pcoa_relab_data <- 
  tibble(
    sampleID = rownames(pcoa_relab$points),
    PCoA = pcoa_relab$points
  ) |> 
  left_join(
    SE_mg |> colData() |> as_tibble(),
    by = join_by(sampleID)
  )

```


#### Type of sample

```{r}

sample_type_colors <- c("gray", "dodgerblue", "orange", "black", "red")

pcoa_plot(pcoa_relab_data, pcoa_relab, "SampleType", c(1, 2)) + scale_color_manual(values = sample_type_colors) 
pcoa_plot(pcoa_relab_data, pcoa_relab, "SampleType", c(3, 4)) + scale_color_manual(values = sample_type_colors) 

```

```{r}

manova <- adonis2(dist_matrix_relab ~ pcoa_relab_data$SampleType)
manova

```


#### Extraction Plate


```{r}

pcoa_plot(pcoa_relab_data, pcoa_relab, "Ext_Lib_Plate", c(1, 2))
pcoa_plot(pcoa_relab_data, pcoa_relab, "Ext_Lib_Plate", c(3, 4))

```


Permutation test 

```{r}

manova <- adonis2(dist_matrix_relab ~ pcoa_relab_data$Ext_Lib_Plate)
manova

```


#### Selected for re-extraction

```{r}

pcoa_plot(pcoa_relab_data, pcoa_relab, "Selected4re.extraction", c(1, 2)) 
pcoa_plot(pcoa_relab_data, pcoa_relab, "Selected4re.extraction", c(3, 4)) 

```


#### Library Pool

```{r}

pcoa_plot(pcoa_relab_data, pcoa_relab, "Library.Pool", c(1, 2))
pcoa_plot(pcoa_relab_data, pcoa_relab, "Library.Pool", c(3, 4))

```


#### Lane


```{r}

pcoa_plot(pcoa_relab_data |> filter(Lane %in% 1:4), pcoa_relab, "Lane", c(1, 2))
pcoa_plot(pcoa_relab_data |> filter(Lane %in% 1:4), pcoa_relab, "Lane", c(3, 4))

```



#### Visit


```{r}

pcoa_plot(pcoa_relab_data |> filter(SampleType == "ClinicalSample", VisitCode %in% c(seq(1000, 1900, by =100), 2120)), pcoa_relab, "VisitCode", c(1, 2)) +
  scale_color_manual(values = c("red", "green3", colorRampPalette(c("dodgerblue1", "dodgerblue4","purple"))(7))) 
pcoa_plot(pcoa_relab_data |> filter(SampleType == "ClinicalSample", VisitCode %in% c(seq(1000, 1900, by =100), 2120)), pcoa_relab, "VisitCode", c(3, 4)) +
  scale_color_manual(values = c("red", "green3", colorRampPalette(c("dodgerblue1", "dodgerblue4","purple"))(7))) 

```




## Saving `SummarizedExperiment` objects

We save the `SE` objects to disk

```{r}

saveRDS(
  SE_mg, 
  str_c(
    get_output_dir(data_source = data_source),  
    "02_se_mg_", today() |> str_remove_all("-"), ".rds"
    )
  )

```



