---
title: "Assembling assays into a master `MultiAssayExperiment` object"
author: Laura Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: refresh # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---

```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(gt)
library(patchwork)
library(SummarizedExperiment)
library(tidySummarizedExperiment)
library(MultiAssayExperiment)

tmp <- fs::dir_map("R scripts/", source)

theme_set(theme_light())

```


# Clinical data (CRF data)

> TODO: fix the path to the VIBRANT dropbox once CRF data cleaning and QCed is finalized

```{r}

crf_data_path <- fs::dir_ls(
  "/Users/laurasymul/OneDrive - UCL/Academia/Research/VIBRANT clinical data UCLouvain/Data processed/"
) |> sort(decreasing = TRUE) |> str_subset("2025-05-26") 

```

## CRF data (raw and QCed)

```{r}

load(str_c(crf_data_path, "/crf_raw.RData"), verbose = TRUE)
load(str_c(crf_data_path, "/crf_clean.RData"), verbose = TRUE)
load(str_c(crf_data_path, "/crf_plates_dictionary.RData"), verbose = TRUE)

```


## `Participants` table

```{r}

load(str_c(crf_data_path, "/participants.RData"), verbose = TRUE)
load(str_c(crf_data_path, "/participants_variable_dictionary.RData"), verbose = TRUE)

```

## `Visits` table

```{r}

load(str_c(crf_data_path, "/visits.RData"), verbose = TRUE)

```


```{r}
#| eval: false
#| fig-width: 25

visits |> 
  ggplot() +
  aes(x = study_day, y = pid |> factor(), color = dfseq |> factor()) +
  geom_point() 

```

We define the `uid` (unique identifier):


```{r}

visits <- 
  visits |> 
  mutate(
    uid = str_c(pid, visit_code, sep = "_")
  ) |> 
  select(uid, everything())


```




# Assays: loading the `SummarizedExperiment` objects 

> TODO: a function that checks the "conformity" of the `SE` objects (e.g. `uid`, `colnames`, `sample_type`, `control_type`, `crf_match`) 

```{r}

data_source <- "real"
data_dir <- str_c(get_data_dir(data_source), "01 Preprocessed and QCed/")
se_files <- fs::dir_ls(data_dir, regexp = "se_.*\\.rds$") |> sort(decreasing = TRUE)

```

## Metagenomics data

```{r}

mg_file <- se_files[grepl("mg", se_files)][1]
se_mg <- readRDS(mg_file)
se_mg

```

## qPCR data

```{r}

qPCR_file <- se_files[grepl("pcr_agg", se_files)][1]
se_qPRC <- readRDS(qPCR_file)
se_qPRC

```
## 16S rRNA amplicon sequencing


```{r}

amplicon_file <- se_files[grepl("16S_agg", se_files)][1]
se_16S <- readRDS(amplicon_file)
se_16S

```


## Luminex data


```{r}

luminex_file <- se_files[grepl("luminex_agg", se_files)][1]
se_luminex <- readRDS(luminex_file)
se_luminex

```

# Creating the `MultiAssayExperiment` object

## Assay list

```{r}

assay_list <- 
  bind_rows(
    tibble(
      mae_assay_name = "mg", 
      se_object_name = "se_mg", 
      assay_label = "Metagenomic"
      ),
    tibble(
      mae_assay_name = "qPCR", 
      se_object_name = "se_qPRC", 
      assay_label = "qPCR"
      ),
    tibble(
      mae_assay_name = "amplicon", 
      se_object_name = "se_16S", 
      assay_label = "16S rRNA amplicon seq."
      ),
    tibble(
      mae_assay_name = "luminex", 
      se_object_name = "se_luminex", 
      assay_label = "Luminex"
      )
  ) |> 
  mutate(
    assay_label = assay_label |> fct_inorder()
  )

```

We will assemble the following tables (`SE` objects):

```{r}
assay_list |> gt()
```

> TODO: add the Flow Cytometry data once available


## `mae_coldata_assays` tables


```{r}

available_samples <- 
  map(
    1:nrow(assay_list),
    function(i, assay_list){
      se <- get(assay_list$se_object_name[i])
      tibble(
        assay = assay_list$mae_assay_name[i],
        se_colname = se |> colnames(), 
        uid = se$uid, 
        sample_type = se$sample_type,
        control_type = se$control_type,
        )
    },
    assay_list = assay_list
  ) |> 
  bind_rows()

if (!all(available_samples$se_colname == available_samples$uid, na.rm = TRUE)) {
  stop("The uid column in the SE objects does not match their colnames")
}

mae_coldata_assays <-
  available_samples |>
  group_by(uid) |>
  summarize(
    n_sample_type = n_distinct(sample_type),
    n_control_type = n_distinct(control_type),
    sample_type = sample_type[1],
    control_type = control_type[1],
    assays = str_c(assay, collapse = ", ")
  )

if (any(mae_coldata_assays$n_sample_type > 1)) {
  warning("Some samples have multiple sample types")
}

if (any(mae_coldata_assays$n_control_type > 1)) {
  warning("Some samples have multiple control types")
}

mae_coldata_assays <- 
  mae_coldata_assays |> 
  select(uid, sample_type, control_type, assays) 


```


## `MAE@colData`: joining the `visits` and `mae_coldata_assays` tables


```{r}

mae_coldata <- 
  # we first do a full join of the CRF visits and mae_coldata_assays by uids;
  dplyr::full_join(
    visits |> select(-starts_with("specimen_collection")) |> mutate(in_CRF = TRUE),
    mae_coldata_assays |> mutate(in_assays = TRUE) |> select(in_assays, everything()),
    by = join_by(uid)
  ) |> 
  mutate(
    in_CRF = in_CRF |> replace_na(FALSE),
    in_assays = in_assays |> replace_na(FALSE)
  )

# we add the participant-invariant information from the `participants` table
mae_coldata <- 
  mae_coldata |> 
  dplyr::left_join(
    participants,
    by = join_by(pid)
  )

```

```{r}
#| eval: true

mae_coldata |> 
  dplyr::count(in_CRF, randomized, in_assays, sample_type, assays) |> 
  gt()

```


```{r}

mae_coldata <- 
  mae_coldata |> 
  mutate(
    study_week = 
      case_when(
        visit_code %in% c(seq(1000, 1500, by = 100), "1700", "1900", "2120") ~ 
          visit_code |> str_sub(2,3) |> str_remove("0") |> as.integer(),
        TRUE ~ NA
      )
  ) |> 
  relocate(study_week, .after = "study_day")

```



## Creating an assay to document specimen collection, and available data


- `specimen_collection_swab` (`collected`, `not collected`, `unclear`)
- `specimen_collection_softcup` (`collected`, `not collected`, `unclear`)
- `specimen_collection_cytobrush` (`collected`, `not collected`, `unclear`)
- `mg` 
- `qPCR`
- `amplicon`
- `luminex`
- `flow`

```{r}

sample_and_visit_info_assay <- 
  mae_coldata |> 
  select(uid, sample_type, in_CRF, in_assays, randomized) 

sample_and_visit_info_assay <- 
  sample_and_visit_info_assay |> 
  dplyr::left_join(
    visits |> select(uid, starts_with("specimen_collection")),
    by = join_by(uid)
  ) 

sample_and_visit_info_assay <-
  sample_and_visit_info_assay |>
  expand_grid(assay = assay_list$mae_assay_name) |>
  dplyr::left_join(
    available_samples |>
      select(uid, assay) |>
      mutate(value = TRUE),
    by = join_by(uid, assay)
  ) |>
  mutate(value = value |> replace_na(FALSE)) |>
  pivot_wider(names_from = assay, values_from = value)

# sample_and_visit_info_assay <- 
#   sample_and_visit_info_assay |> 
#   mutate(
#     mg_category = 
#       case_when(
#         !str_detect(sample_type, "Clinical sample") & mg ~ "Not a clinical sample",
#         is.na(randomized) & mg ~ "???",
#         !is.na(randomized) & !randomized & mg ~ "Additional sample from non-randomized participants",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & mg ~ "Expected data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & mg ~ "Unexpected data",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & !mg ~ "Missing data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & !mg ~ "Sample not collected",
#         TRUE ~ "ERROR"
#       ),
#     qPCR_category = 
#       case_when(
#         !str_detect(sample_type, "Clinical sample") & qPCR ~ "Not a clinical sample",
#         is.na(randomized) & qPCR ~ "???",
#         !is.na(randomized) & !randomized & qPCR ~ "Additional sample from non-randomized participants",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & qPCR ~ "Expected data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & qPCR ~ "Unexpected data",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & !qPCR ~ "Missing data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & !qPCR ~ "Sample not collected",
#         TRUE ~ "ERROR"
#       ),
#     amplicon_category = 
#       case_when(
#         !str_detect(sample_type, "Clinical sample") & amplicon ~ "Not a clinical sample",
#         is.na(randomized) & amplicon ~ "???",
#         !is.na(randomized) & !randomized & amplicon ~ "Additional sample from non-randomized participants",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & amplicon ~ "Expected data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & amplicon ~ "Unexpected data",
#         !is.na(specimen_collection_swab) & (specimen_collection_swab > 0) & !amplicon ~ "Missing data",
#         (is.na(specimen_collection_swab) | (specimen_collection_swab == 0)) & !amplicon ~ "Sample not collected",
#         TRUE ~ "ERROR"
#       ),
#     luminex_category = 
#       case_when(
#         !str_detect(sample_type, "Clinical sample") & luminex ~ "Not a clinical sample",
#         is.na(randomized) & luminex ~ "???",
#         !is.na(randomized) & !randomized & luminex ~ "Additional sample from non-randomized participants",
#         !is.na(specimen_collection_softcup) & (specimen_collection_softcup != "not collected") & luminex ~ "Expected data",
#         (is.na(specimen_collection_softcup) | (specimen_collection_softcup == "not collected")) & luminex ~ "Unexpected data",
#         !is.na(specimen_collection_softcup) & (specimen_collection_softcup == "collected") & !luminex ~ "Missing data",
#         (is.na(specimen_collection_softcup) | (specimen_collection_softcup == "not collected")) & !luminex ~ "Sample not collected",
#         TRUE ~ "ERROR"
#       )
#   )

# 
# The last 5 columns can take the following values:
# 
# - `expected` (swab collected and available data), 
# - `unexpected` (swab not collected but available data), 
# - `missing` (swab collected but no available data),
# - `maybe missing` (unclear if swab was collected and no available data),
# - `not collected` (swab not collected), 
# - `additional` (swab collected from a non-randomized participant), 
# - `NA` (not collected, no available data)
# 


```

```{r}

se_sample_and_visit_info_assay <- 
  SummarizedExperiment(
    assays = list(
      sample_and_visit_info = 
        sample_and_visit_info_assay |> 
        select(-c(sample_type, in_CRF, in_assays, randomized)) |> 
        as.data.frame() |> 
        column_to_rownames("uid") |> 
        t()
    ),
    colData = 
      sample_and_visit_info_assay |> select(uid) |> 
      mutate(rownames = uid) |> as.data.frame() |> 
      column_to_rownames("rownames")
  )
```




## Creating the `MultiAssayExperiment` object


```{r}

SE_list <- 
  list(
    visit_and_sample_info = se_sample_and_visit_info_assay,
    mg = se_mg,
    qPCR = se_qPRC,
    amplicon = se_16S,
    luminex = se_luminex
  )


mae <- 
  MultiAssayExperiment(
    experiments = MultiAssayExperiment::ExperimentList(SE_list),
    colData = 
      mae_coldata |> select(-in_CRF, -in_assays) |> 
      as.data.frame() |> mutate(rownames = uid) |> 
      column_to_rownames("rownames"),
    metadata = list(
      study = "VIBRANT",
      data_source = "actual study data (not simulated)",
      data_status = "partially QCed",
      date = today(),
      colData_dictionary = participants_variable_dictionary,
      crf_plates_description = crf_plates_dictionary,
      crf_data_raw = crf_raw,
      crf_data_clean = crf_clean
    )
  )

```


```{r}
mae
```




# Available assays

```{r}

mae_sample_map <- 
  mae@sampleMap |> as_tibble() |> filter(assay %in% assay_list$mae_assay_name) |>
  select(uid = primary, assay) |> 
  arrange(uid) |> 
  dplyr::left_join(assay_list |> dplyr::rename(assay = mae_assay_name), by = join_by(assay)) |> 
  dplyr::right_join(
    mae@colData |> as.data.frame() |> as_tibble() |> 
      select(uid, pid, visit_code, visit_type, sample_type, location, randomized, arm), 
    by = join_by(uid)
    )

```

All samples

```{r}
#| fig-height: 4
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  group_by(uid) |> mutate(n = n()) |> ungroup() |> 
  mutate(uid = uid |> fct_reorder(-n)) |>
  ggplot() +
  aes(x = uid, y = assay_label |> fct_rev(), fill = assay_label) +
  geom_tile() +
  facet_grid(. ~ sample_type, scales = "free_x", space = "free_x") +
  scale_x_discrete("Sample IDs", breaks = NULL) +
  ylab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```


Longitudinal availability for all visits


```{r}
#| fig-height: 12
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  # filter(visit_code %in% c("0000", seq(1000, 1900, by = 100), "2120")) |> 
  ggplot() +
  aes(x = assay_label, y = pid, fill = assay_label) +
  geom_tile() +
  facet_grid(ifelse(randomized, "Randomized", "Not randomized") + location ~ visit_code, scales = "free_y", space = "free_y") + # sample_category
  scale_y_discrete("Participants IDs") + #, breaks = "NULL")  +
  xlab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_blank(), #element_text(angle = 45, hjust = 1),
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```
For the weekly visits

```{r}
#| fig-height: 12
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  filter(visit_type == "Clinic") |> 
  ggplot() +
  aes(x = assay_label, y = pid, fill = assay_label) +
  geom_tile() +
  facet_grid(
    location + ifelse(randomized, "Randomized", "Not randomized")  + arm  ~ visit_code, 
    scales = "free_y", space = "free_y"
    ) + # sample_category
  scale_y_discrete("Participants IDs") + #, breaks = "NULL")  +
  xlab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_blank(), #element_text(angle = 45, hjust = 1),
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```

> There is one US participant with the termination `visit_code` for their 16S data, but it should be changed to `1700`.


# Subsetting and saving the `MultiAssayExperiment` objects

We save the "master" MAE (with all assays); but if needed for publication, we can subset the MAE to only include the assays/data of interest.

```{r}

saveRDS(
  mae, 
  str_c(
    get_data_dir(data_source = data_source),  
    "02 MAEs/",
    "mae_full_", today() |> str_remove_all("-"), ".rds"
  )
)

```



