---
title: "Assembling assays into a master `MultiAssayExperiment` object"
author: Laura Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: true # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---

```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(gt)
library(patchwork)
library(SummarizedExperiment)
library(tidySummarizedExperiment)
library(MultiAssayExperiment)

tmp <- fs::dir_map("R scripts/", source)

theme_set(theme_light())

```


# Clinical data (CRF data)

> TODO: fix the path to the VIBRANT dropbox once CRF data cleaning and QCed is finalized

```{r}

crf_data_path <- fs::dir_ls(
  "/Users/laurasymul/OneDrive - UCL/Academia/Research/VIBRANT clinical data UCLouvain/Data processed/"
) |> sort(decreasing = TRUE) |> str_subset("2025-05-20") 


```



## `Participants` table

```{r}

load(str_c(crf_data_path, "/participants.RData"), verbose = TRUE)

participants <- 
  participants |> 
  mutate(randomized = !is.na(treatment_code)) |> 
  select(pid, randomized, group4, reason_randomisation, reason_termination, reason_replacement, everything())

```

## `Visits` table

```{r}

load(str_c(crf_data_path, "/visits.RData"), verbose = TRUE)
visits_long <- visits
rm(visits)

```

```{r}
#| eval: false
#| fig-width: 25

visits_long |> 
  ggplot() +
  aes(x = study_day, y = pid |> factor(), color = crf_name) +
  geom_point() +
  facet_grid(. ~ dfseq, scales = "free", space = "free")

```

```{r}

visits <- 
  visits_long |> 
  group_by(pid, dfseq) |> 
  summarize(study_day = median(study_day), .groups = "drop")

```


```{r}
#| eval: false
#| fig-width: 25

visits |> 
  filter(study_day > -100) |> 
  ggplot() +
  aes(x = study_day, y = pid |> factor(), color = dfseq |> factor()) +
  geom_point() 

```

## `population flags` table

> TODO


## `MAE@colData`: merging these tables to create the cross-assay sample and participant information table

```{r}

mae_coldata_crf <- 
  participants |> 
  mutate(
    location = ifelse(str_detect(pid |> as.character(), "^10"), "US", "SA"),
    pid = str_c("068", pid),
  ) |> 
  select(pid, location, randomized, group4, reason_randomisation, reason_termination, reason_replacement) |>
  dplyr::left_join(
    visits |> 
      mutate( pid = str_c("068", pid)) |> 
      select(pid, dfseq, study_day) |> 
      mutate(visit_code = dfseq |> str_pad(width = 4, pad = "0")),
    by = join_by(pid)
  ) |> 
  mutate(
    uid = str_c(pid, "_", visit_code)
  ) |> 
  select(uid, everything())

```


# Assays: loading the `SummarizedExperiment` objects 

> TODO: a function that checks the "conformity" of the `SE` objects (e.g. `uid`, `colnames`, `sample_type`, `control_type`, `crf_match`) 

```{r}

data_source <- "real"
data_dir <- str_c(get_data_dir(data_source), "01 Preprocessed and QCed/")
se_files <- fs::dir_ls(data_dir, regexp = "se_.*\\.rds$") |> sort(decreasing = TRUE)

```

## Metagenomics data

```{r}

mg_file <- se_files[grepl("mg", se_files)][1]
se_mg <- readRDS(mg_file)
se_mg

```

## qPCR data

```{r}

qPCR_file <- se_files[grepl("pcr_agg", se_files)][1]
se_qPRC <- readRDS(qPCR_file)
se_qPRC

```
## 16S rRNA amplicon sequencing


```{r}

amplicon_file <- se_files[grepl("16S_agg", se_files)][1]
se_16S <- readRDS(amplicon_file)
se_16S

```


## Luminex data


```{r}

luminex_file <- se_files[grepl("luminex_agg", se_files)][1]
se_luminex <- readRDS(luminex_file)
se_luminex

```

# Creating the `MultiAssayExperiment` object

## Assay list

```{r}

assay_list <- 
  bind_rows(
    tibble(
      mae_assay_name = "mg", 
      se_object_name = "se_mg", 
      assay_label = "Metagenomic"
      ),
    tibble(
      mae_assay_name = "qPCR", 
      se_object_name = "se_qPRC", 
      assay_label = "qPCR"
      ),
    tibble(
      mae_assay_name = "amplicon", 
      se_object_name = "se_16S", 
      assay_label = "16S rRNA amplicon seq."
      ),
    tibble(
      mae_assay_name = "luminex", 
      se_object_name = "se_luminex", 
      assay_label = "Luminex"
      )
  ) |> 
  mutate(
    assay_label = assay_label |> fct_inorder()
  )

```

We will assemble the following tables (`SE` objects):

```{r}
assay_list |> gt()
```




## `mae_coldata_assays` tables


```{r}

available_samples <- 
  map(
    1:nrow(assay_list),
    function(i, assay_list){
      se <- get(assay_list$se_object_name[i])
      tibble(
        assay = assay_list$mae_assay_name[i],
        se_colname = se |> colnames(), 
        uid = se$uid, 
        sample_type = se$sample_type,
        control_type = se$control_type,
        )
    },
    assay_list = assay_list
  ) |> 
  bind_rows()

if (!all(available_samples$se_colname == available_samples$uid, na.rm = TRUE)) {
  stop("The uid column in the SE objects does not match their colnames")
}

mae_coldata_assays <-
  available_samples |>
  group_by(uid) |>
  summarize(
    n_sample_type = n_distinct(sample_type),
    n_control_type = n_distinct(control_type),
    sample_type = sample_type[1],
    control_type = control_type[1],
    assays = str_c(assay, collapse = ", ")
  )

if (any(mae_coldata_assays$n_sample_type > 1)) {
  warning("Some samples have multiple sample types")
}

if (any(mae_coldata_assays$n_control_type > 1)) {
  warning("Some samples have multiple control types")
}

mae_coldata_assays <- 
  mae_coldata_assays |> 
  select(uid, sample_type, control_type, assays) 


```


## `MAE@colData`: joining the `mae_coldata_crf` and `mae_coldata_assays` tables

```{r}

mae_coldata <- 
  dplyr::full_join(
    mae_coldata_crf |> select(uid, pid, visit_code, study_day) |> mutate(in_CRF = TRUE),
    mae_coldata_assays |> mutate(in_assays = TRUE),
    by = join_by(uid)
  ) |> 
  mutate(
    in_CRF = in_CRF |> replace_na(FALSE),
    in_assays = in_assays |> replace_na(FALSE)
  ) |> 
  dplyr::left_join(
    mae_coldata_crf |> 
      select(
        pid, location, randomized, group4, 
        reason_randomisation, reason_termination, reason_replacement
        ) |> 
      distinct(),
    by = join_by(pid)
  )

```

```{r}
#| eval: false

mae_coldata |> 
  dplyr::count(in_CRF, randomized, in_assays, sample_type, assays) |> 
  gt()

```

## Creating an assay to document visit attendance, specimen collection, and available data

> TODO: add visit attendance, specimen collection
> will have the following columns:

- `visit_type` (`clinic` or `home`)
- `visit_category` (`planned`, `additional`)
- `visit_attendance` (`attended`, `missed`)
- `specimen_collection_swab` (`collected`, `not collected`, `unclear`)
- `specimen_collection_softcup` (`collected`, `not collected`, `unclear`)
- `specimen_collection_cytobrush` (`collected`, `not collected`, `unclear`)
- `mg` 
- `qPCR`
- `amplicon`
- `luminex`
- `flow`

The last 5 columns can take the following values:

- `expected` (swab collected and available data), 
- `unexpected` (swab not collected but available data), 
- `missing` (swab collected but no available data),
- `maybe missing` (unclear if swab was collected and no available data),
- `not collected` (swab not collected), 
- `additional` (swab collected from a non-randomized participant), 
- `NA` (not collected, no available data)


> For now, we'll only have the last 5 columns and they'll be binary (we have data or not)

```{r}

sample_and_visit_info_assay <- 
  mae_coldata |> 
  select(uid, in_CRF, in_assays)

sample_and_visit_info_assay <- 
  sample_and_visit_info_assay |> 
  expand_grid(assay = assay_list$mae_assay_name) |> 
  dplyr::left_join(
    available_samples |>
      select(uid, assay) |> 
      mutate(value = TRUE),
    by = join_by(uid, assay)
  ) |> 
  mutate(value = value |> replace_na(FALSE)) |> 
  pivot_wider(names_from = assay, values_from = value) 

se_sample_and_visit_info_assay <- 
  SummarizedExperiment(
    assays = list(
      sample_and_visit_info = 
        sample_and_visit_info_assay |> 
        select(-in_assays) |> 
        as.data.frame() |> 
        column_to_rownames("uid") |> 
        t()
    ),
    colData = 
      sample_and_visit_info_assay |> select(uid) |> 
      mutate(rownames = uid) |> as.data.frame() |> 
      column_to_rownames("rownames")
  )

```



## Creating the `MultiAssayExperiment` object


```{r}

SE_list <- 
  list(
    visit_and_sample_info = se_sample_and_visit_info_assay,
    mg = se_mg,
    qPCR = se_qPRC,
    amplicon = se_16S,
    luminex = se_luminex
  )


mae <- 
  MultiAssayExperiment(
    experiments = MultiAssayExperiment::ExperimentList(SE_list),
    colData = 
      mae_coldata |> select(-in_CRF, -in_assays) |> 
      as.data.frame() |> mutate(rownames = uid) |> 
      column_to_rownames("rownames"),
    metadata = list(
      study = "VIBRANT",
      MAE_creation_date = today()
    )
  )

```


```{r}
mae
```




# Available assays

```{r}

mae_sample_map <- 
  mae@sampleMap |> as_tibble() |> filter(assay %in% assay_list$mae_assay_name) |>
  select(uid = primary, assay) |> 
  arrange(uid) |> 
  dplyr::left_join(assay_list |> dplyr::rename(assay = mae_assay_name), by = join_by(assay)) |> 
  dplyr::right_join(
    mae@colData |> as.data.frame() |> as_tibble() |> 
      select(uid, pid, visit_code, sample_type, location, randomized, group4), 
    by = join_by(uid)
    )

```

All samples

```{r}
#| fig-height: 4
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  group_by(uid) |> mutate(n = n()) |> ungroup() |> 
  mutate(uid = uid |> fct_reorder(-n)) |>
  ggplot() +
  aes(x = uid, y = assay_label |> fct_rev(), fill = assay_label) +
  geom_tile() +
  facet_grid(. ~ sample_type, scales = "free_x", space = "free_x") +
  scale_x_discrete("Sample IDs", breaks = NULL) +
  ylab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```


Longitudinal availability for all visits


```{r}
#| fig-height: 12
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  # filter(visit_code %in% c("0000", seq(1000, 1900, by = 100), "2120")) |> 
  ggplot() +
  aes(x = assay_label, y = pid, fill = assay_label) +
  geom_tile() +
  facet_grid(ifelse(randomized, "Randomized", "Not randomized") + location ~ visit_code, scales = "free_y", space = "free_y") + # sample_category
  scale_y_discrete("Participants IDs") + #, breaks = "NULL")  +
  xlab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_blank(), #element_text(angle = 45, hjust = 1),
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```
For the weekly visits

```{r}
#| fig-height: 12
#| fig-width: 12

mae_sample_map |> 
  filter(!is.na(assay)) |> 
  filter(visit_code %in% c("0000", seq(1000, 1900, by = 100), "2120")) |> 
  ggplot() +
  aes(x = assay_label, y = pid, fill = assay_label) +
  geom_tile() +
  facet_grid(
    location + 
      ifelse(randomized, "Randomized", "Not randomized") + 
      ifelse(group4 == "Yes", "Overlap", "Other arm") ~ 
      visit_code, 
    scales = "free_y", space = "free_y"
    ) + # sample_category
  scale_y_discrete("Participants IDs") + #, breaks = "NULL")  +
  xlab("") +
  scale_fill_brewer("Assay", type = "qual", palette = 7) + # 3
  theme(
    axis.text.x = element_blank(), #element_text(angle = 45, hjust = 1),
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90),
    legend.position = "bottom"
    )

```


# Subsetting and saving the `MultiAssayExperiment` objects

We save the "master" MAE (with all assays); but if needed for publication, we can subset the MAE to only include the assays/data of interest.

```{r}

saveRDS(
  mae, 
  str_c(
    get_data_dir(data_source = data_source),  
    "02 MAEs/",
    "mae_full_", today() |> str_remove_all("-"), ".rds"
  )
)

```



